README
Mojolicious-Plugin-Prototype-MooseForm

The README is used to introduce the module and provide instructions on
how to install the module, any machine dependencies it may have (for
example C compilers and installed libraries) and any other information
that should be provided before the module is installed.

A README file is required for CPAN modules since CPAN extracts the README
file from a module distribution so that people browsing the archive
can use it to get an idea of the module's uses. It is usually a good idea
to provide version information here so that people can decide whether
fixes for the module are worth downloading.


INSTALLATION

To install this module, run the following commands:

	perl Makefile.PL
	make
	make test
	make install

SUPPORT AND DOCUMENTATION

After installing, you can find documentation for this module with the
perldoc command.

    perldoc Mojolicious::Plugin::Prototype::MooseForm

You can also look for information at:

    RT, CPAN's request tracker (report bugs here)
        http://rt.cpan.org/NoAuth/Bugs.html?Dist=Mojolicious-Plugin-Prototype-MooseForm

    AnnoCPAN, Annotated CPAN documentation
        http://annocpan.org/dist/Mojolicious-Plugin-Prototype-MooseForm

    CPAN Ratings
        http://cpanratings.perl.org/d/Mojolicious-Plugin-Prototype-MooseForm

    Search CPAN
        http://search.cpan.org/dist/Mojolicious-Plugin-Prototype-MooseForm/


LICENSE AND COPYRIGHT

Copyright (C) 2012 Fernando Correa de Oliveira

This program is free software; you can redistribute it and/or modify it
under the terms of either: the GNU General Public License as published
by the Free Software Foundation; or the Artistic License.

See http://dev.perl.org/licenses/ for more information.



Changes
Revision history for Mojolicious-Plugin-Prototype-MooseForm

0.01    Date/time
        First version, released on an unsuspecting world.



t/app/template.pl
#!/usr/bin/perl 
package bla;
use Moose;
use Moose::Util::TypeConstraints;
has a => (is => 'ro', isa => "Num");
has e => (is => 'ro', default => "Bla");
has i => (is => 'ro', isa => "ArrayRef[Str]", documentation => "Doc for attr 'i'");
has o => (is => 'ro', required => 1);
has u => (is => 'ro', isa => "Maybe[Num]", documentation => "Doc for attr 'u'");
has x => (is => 'rw', isa => "Bool");
has y => (is => 'rw', isa => enum([qw/ la le li lo lu/]));

package main;
use Mojolicious::Lite;

BEGIN{ plugin "Mojolicious::Plugin::Prototype::MooseForm"; }

app->prototype_conf("required_symbol" => "<b>Obrigat&oacute;rio</b>");

#get "/"  => sub{shift()->get_defaults("bla")} => "moose_form";
post "/" => sub{
   my $self = shift;
   my $obj = $self->create_object("bla");
   return $self->redirect_to("/mf") if not $obj;
   $self->render_json({ %$obj });
};

get_moose_form "/mf" => bla => {action => "/"} => "moose_form";

moose_form "/done" => "bla";

app->start;

__DATA__

@@ done.html.ep
% use Data::Dumper;
<pre>
% for my $key(sort keys %$obj) {
  <%= "$key: " . Dumper $obj->{ $key } =%><br>
%}
</pre>


t/lib/test_extension.pm
package test_extension;
use Moose;
extends "Mojolicious::Plugin::Prototype::MooseForm::Templater";

__DATA__

@@ _test.bli.ep
from extension
@@ _test.blu.ep
from extension


lib/Mojolicious/Plugin/Prototype/MooseForm/TemplateData.pm
package Mojolicious::Plugin::Prototype::MooseForm::TemplateData;
use Moose;
42

__DATA__

@@ moose_form_template_say_required_default.html.ep
<span class=required></span>

@@ moose_form_template_say_required_1.html.ep
<span class=required><%= b(prototype_conf("required_symbol")) =%></span>

@@ moose_form_template_change_type___anon__.html.ep
<span class="radio_container">
   <% for my $opt(@{ $attr->{ possibilities } }) { %>
      <input
       type="radio"
       name="<%= $attr->{name} =%>"
       value="<%= $opt =%>"
       <% if(defined $attr->{value} and $attr->{value} eq $opt) { %>
          checked=1
       <% } %>
      >
      <%= $opt =%><BR>
   <% } %>
</span>

@@ moose_form_template_change_type_bool.html.ep
% $$required = 0;
<input
 type="checkbox"
 name="<%= $attr->{name} =%>"
 value="1"
 <% if($attr->{value}) { =%>
    checked=1
 <% } =%>
>

@@ moose_form_template_change_type_arrayref.html.ep
% $$required = 0;
% my $array_req = 1;
<div class="array_item_base submit_remove">
   <div class=item>
      <%= include moose_form_template_for( "change", "type", $subtype ), attr => $attr, type => $subtype, required => \$array_req =%>
      <%= include moose_form_template_for( "say", "required", $array_req ), attr => $attr =%>
      <input class=remove type=button value="-"><br>
   </div>
</div>
<div class=array_active>
   % my $counter = 0;
   % my $orig_name = $attr->{ name };
   % for my $val(@{ $attr->{ value } }) {
      % $counter++;
      % $attr->{ value } = $val;
      % $attr->{ name } = $orig_name . $counter;
      <div class=item>
         <%= include moose_form_template_for( "change", "type", $subtype ), attr => $attr, type => $subtype, required => \$array_req =%>
         <%= include moose_form_template_for( "say", "required", $array_req ), attr => $attr =%>
         <input class=remove type=button value="-"><br>
      </div>
   % }
</div>

<input class=add type=button value="+" count="<%= $counter =%>">

@@ moose_form_template_change_type_any.html.ep
% $$required = 0;
<%= include "moose_form_template_change_type_default", attr => $attr, type => $subtype, required => $required =%>

@@ moose_form_template_change_type_maybe.html.ep
% $$required = 0;
<%= include moose_form_template_for( "change", "type", $subtype ), attr => $attr, type => $subtype, required => $required =%>

@@ moose_form_template_change_type_default.html.ep
<input
 class="attr_input type_<%= join " ", lc($type ? $type : $attr->{ type }), ($$required ? "inp_required" : ()), (exists $error->{$attr->{name}} ? "input_error" : ()) =%>"
 type="text"
 name="<%= $attr->{ name }  =%>"
 value="<%= $attr->{ value }  =%>"
>

@@ moose_form.html.ep
<link rel="stylesheet" type="text/css" href="<%= url_for "/css/moose_form.css" =%>" />
<script src="<%= url_for "js/jquery.js" =%>"></script>
<script src="<%= url_for "js/main.js" =%>"></script>
<form class="moose_form_<%= $class =%>" method=POST action="<%= $action =%>?rand=<%= rand =%>">
   <% if($attributes) { =%>
      <%= include "moose_form_table", attributes => $attributes =%>
   <% } =%>
</form>

@@ moose_form_table.html.ep
% my $num_of_colors = scalar @{ moose_form_get_conf()->{prototype_list_bgcolor} };
<table class=moose_form_table>
% my $counter = 0;
% for my $attr(sort {$a->{ name } cmp $b->{ name }} @$attributes) {
   <tr class="bgcolor_<%= ( $counter++ % $num_of_colors ) + 1 %>">
   %= include "moose_form_line", attr => $attr;
   </tr>
% }
   <tr class="bgcolor_<%= ( $counter++ % $num_of_colors ) + 1 %>">
   %= include "moose_form_last_line";
   </tr>
</table>

@@ moose_form_last_line.html.ep
% my $value = moose_form_get_conf()->{prototype_submit_label};
<td colspan=2 align=right>
   <input type=submit value="<%= $value =%>"
</td>

@@ moose_form_line.html.ep
% my $required = 1;
<td>
   <%= include moose_form_template_for( "title", "none", "bla" ), attr => $attr =%>
</td>
<td>
   <%= include moose_form_template_for( "change", "type", $attr->{ type } ), attr => $attr, type => $attr->{  type } , required => \$required =%>
   <%= include moose_form_template_for( "say", "required", $required ), attr => $attr =%>
   <div
    class=error_msg
    <% if($error->{$attr->{name}}) { =%>
       style="display: block;"
    <% } =%>
   >
      <% if( $attr->{ doc } ) { =%>
         <%= $attr->{ doc } =%>
      <% } else { =%>
         <%= moose_form_get_conf()->{prototype_default_error_msg} =%>
      <% } =%>
      <% if($error->{$attr->{name}}) { =%>
         <BR>
         <strong>ERROR:</strong> <%= $error->{$attr->{name}} =%>
      <% } =%>
      </div>
</td>

@@ moose_form_template_title_none_default.html.ep
<span class="attr_line_header"><%= $attr->{ title } =%></span>

@@ moose_form.css.ep
% my $conf = moose_form_get_conf();
% my @colors = @{ $conf->{prototype_list_bgcolor} };
% for( my $index = 0; $index < @colors; $index++) { 
.bgcolor_<%= $index + 1 =%> {
   background-color: <%= $colors[ $index ] =%>;
} 
% }
.moose_form_table {
  width: 100%;
}
.required {
   color: <%= $conf->{ prototype_required_color } =%>;
}
.error_msg {
   border:
      <%= $conf->{ prototype_error_border_width } =%>
      solid
      <%= $conf->{ prototype_error_border_color } =%>;
   background-color: <%= $conf->{ prototype_error_bgcolor } =%>;
   padding: 3px;
   display: none;
}
input.input_error {
   color: <%= $conf->{ prototype_input_error_color } =%>; 
}
div.array_item_base {
   display: none;
}

@@ js/main.js

$(document).ready(function(){
   $(".array_item_base input[type!='button']").val( "" );
   $(".attr_input:text").focus(function(){
      $(this).select();
   });
   $(".attr_input").each(function(){
      this.array_test = [];
      this.gotwrong = function(){ 
         $(this).parents("td").find(".error_msg").show("slow");
         $(this).focus();
         $(this).addClass("input_error");
      };
      this.gotright = function(){ 
         $(this).parents("td").find(".error_msg").hide("slow");
         $(this).removeClass("input_error");
      };
   });
   $(".attr_input").change(function(){
      var array_test = this.array_test;
      var resp = true;
      for(var i = 0; i < array_test.length; i++)
         resp = resp && array_test[i](this);
      if(resp)
         this.gotright();
      else
         this.gotwrong();
   });
   $(".type_num").each(function(){
      this.array_test.push(function(obj){return $(obj).val().match(/^[+-]?\d*(?:.\d+)?$/)});
   });
   $(".inp_required").each(function(){
      this.array_test.push(function(obj){return $(obj).val() != ""});
   });
   $(".add").click(function(){
      $(this).attr("count", parseInt($(this).attr("count")) + 1);
      var new_item = $(this).parents("td").find(".array_item_base div").clone();
      new_item.hide();
      $(new_item).find("input[type!='button']").attr(
         "name", $(new_item).find("input").attr("name") + $(this).attr("count")
      );
      $(this).parents("td").find(".array_active").append( new_item );
      new_item.show("slow", function(){
         new_item.find("input[type!='button']").first().focus();
      });
   });
   $(".array_active input.remove").live("click", function(){
      var iname = $(this).parents(".item").find(".attr_input").attr("name");
      $(this).parents(".item").hide( "slow", function(){ $(this).remove() });
      var add = $(this).parents("td").find(".add");
      add.attr("count", parseInt(add.attr("count")) - 1);

      $(this).parents(".item").nextAll(".item").each(function(){
         var tmpname = $(this).find(".attr_input").attr("name");
         $(this).find(".attr_input").attr("name", iname);
         iname = tmpname;
      });
   });
   $("form.moose_form").submit(function(){
      $(this).find(".submit_remove").remove();
   });
});








lib/Mojolicious/Plugin/Prototype/MooseForm/ClassResolver.pm
package Mojolicious::Plugin::Prototype::MooseForm::ClassResolver;
use Moose;
use v5.10;

has plugin => (is => 'ro', required => 1, handles => [qw/exec can_exec/]);
has conf   => (is => 'ro', required => 1);

sub get_class_details {
   my $self   = shift;
   my $class  = shift;
   my $params = shift || {};
   
   my $meta = $class->meta;

   map {
      my $oval = $params->{$_->name}; 
      my $val = $oval
                ? $oval
                : $_->is_default_a_coderef
                  ? $_->default->($class)
                  : $_->default;
      {
         possibilities => ($_->type_constraint && $_->type_constraint->can("values") ? $_->type_constraint->values : []),
         title         => join(" ", map{ "\u$_" } grep {defined $_} split /_+/, $_->name),
         name          => $_->name,
         value         => $val,
         doc           => $_->documentation,
         type          => $_->type_constraint ? $_->type_constraint . "" : "Any",
         req           => $_->is_required,
      }
   } $meta->get_all_attributes
}

sub attr_error {
   my $self  = shift;
   my $attr  = shift;
   my $value = shift;

   eval{ $attr->verify_against_type_constraint($value) };
   if($@){
      return $1 if (split /\n/, $@)[0] =~ m{^\s*(.+)\s+at\s+.*/\w+\.pm\s+line\s+\d+\s*$};
   }
   return
}

sub create_obj {
   my $self  = shift;
   my $class = shift;
   my $data  = shift;

   my @attrs = $class->meta->get_all_attributes;

   my (%error, %attrs);
   for my $attr(@attrs) {
      my $name  = $attr->name;
      my $type  = $attr->type_constraint || "Any";

      my $value = $self->exec(get_value_for_type => $type, $name, $data);

      $attrs{ $name } = $value;
      my $err;
      $err = $self->exec(attr_error => $attr, $value);
      $error{ $name } = $err if $err;
   }
   if(keys %error) {
      $self->plugin->error({ %error });
      return { %attrs }
   } else {
      return $class->new({ %attrs });
   }
}

sub create_exec_name {
   my $self = shift;
   my $type = join "_of_", map{ lc $_ } @_;


   "get_value_for_" . $type;
}

sub get_value_for_type {
   my $self = shift;
   my $type = lc shift;

   my $ret;
   my @val = $self->exec( separate_value => $type);
   for( reverse 0 .. $#val ) {
      my $name = $self->create_exec_name(@val[ 0 .. $_ ]);
      my $subtype = $self->exec(type_reconstruct => @val[ $_ + 1 .. $#val ]);
      $ret = $self->exec($name => $subtype, @_) if $self->can_exec($name);
   }
   $ret = $self->exec("get_value_for_default" => $type, @_) if not defined $ret;
   $ret || undef
}

sub get_value_for_arrayref {
   my $self = shift;
   my $type = shift;
   my $name = shift;
   my $data = shift;

   my @ret;
   my $count = 1;
   my $new_name = $name . $count++;
   while( exists $data->{ $new_name } ) { 
      my $val = $self->get_value_for_type($type, $new_name, $data, @_);
      push @ret, $val;
      $new_name = $name . $count++;
   }
   [ @ret ]
}
sub get_value_for_maybe {
   my $self = shift;
   return $self->get_value_for_type(@_) || undef
}
sub get_value_for_num   {shift()->get_value_for_default(@_) || undef}
sub get_value_for_str   {shift()->get_value_for_default(@_) || undef}
sub get_value_for_any   {shift()->get_value_for_default(@_) || undef}

sub get_value_for_default {
   my $self = shift;
   my $type = shift;
   my $name = shift;
   my $data = shift;

   $data->{ $name }
}

42


lib/Mojolicious/Plugin/Prototype/MooseForm.pm
package Mojolicious::Plugin::Prototype::MooseForm;
use Mojo::Base 'Mojolicious::Plugin';
use Mojolicious::Plugin::Prototype::MooseForm::ClassResolver;
use Mojolicious::Plugin::Prototype::MooseForm::TemplateData;
use v5.10;
use Moose;
use Carp qw/croak/;

our $VERSION = '0.01';

has conf    => (is => 'rw', lazy => 1, isa => "HashRef", default => sub{{
   prototype_list_bgcolor        => [ "#E6E6FA", "#FFFFFF" ],
   prototype_bgcolor             => "#EEEEEE",
   prototype_submit_label        => "OK",
   prototype_required_color      => "red",
   prototype_required_symbol     => "*",
   prototype_error_bgcolor       => "#ffaaaa",
   prototype_error_border_color  => "red",
   prototype_error_border_width  => "1px",
   prototype_input_error_color   => "red",
   prototype_default_error_msg   => "ERROR: Value don't match the type",

}});
has plugins => (is => 'ro', lazy => 1, default => sub{ [] });
has error   => (is => 'rw', isa => "HashRef");
has app     => (is => 'rw');

after error => sub {
   my $self = shift;
   $self->{ error } = {} if not @_
};

sub plugin_init {
   my $self = shift;
   return {
            plugin => $self,
            conf   => $self->conf,
          }
}

sub set_conf {
   my $self = shift;
   my $conf = shift;

   my $sconf = $self->conf;
   for my $key(keys %$conf) {
     $sconf->{ $key } = $conf->{ $key };
   }
}

sub can_exec {
   my $self    = shift;
   my $exec    = shift;
   my $plugins = $self->plugins;

   for my $plug(@$plugins) {
      return 1 if $plug->can($exec)
   }
   return 1 if $self->can($exec);
   0
}

sub exec {
   my $self    = shift;
   my $exec    = shift;
   my $plugins = $self->plugins;

   for my $plug(@$plugins) {
      if($plug->can($exec)) {
         return $plug->$exec(@_);
      }
   }
   if( $self->can($exec) ) {
      return $self->$exec(@_);
   }
   croak "Can not find '$exec' on plugins."
}

sub add_plugin {
   my $self   = shift;
   my $plugin = shift;

   unshift @{ $self->plugins }, $plugin->new( $self->plugin_init );
   push @{ $self->app->renderer->classes }, $plugin if $self->app && $self->app->can("renderer");
   push @{ $self->app->static->classes }, $plugin if $self->app && $self->app->can("renderer");
}

sub separate_value {
   my $self  = shift;
   my $value = shift;

   return "none" if not defined $value;

   my @ret;
   if(my ($first, $second) = $value =~ /^\s*(\w+)\s*\[\s*(.*?)\s*\]\s*$/) {
      push @ret, $first, $self->separate_value( $second );
   } else {
      push @ret, $value;
   }
   @ret
}

sub create_template_name {
      my $self   = shift;
      my $action = lc shift;
      my $type   = lc shift;
      my $value  = join "_of_", map { lc $_ } @_;

      #$self->app->log->debug("***** TESTING: moose_form_template_${action}_${type}_${value}");
      "moose_form_template_${action}_${type}_${value}"
}

sub type_reconstruct {
   my $self  = shift;
   my $first = shift;

   my $ret;
   if( not $first ) {
   } elsif( not @_ ) {
      $ret = $first;
   } else {
      $ret = "$first\[" . $self->type_reconstruct(@_) . "]"
   }
   $ret
}

sub register { 
   my $self = shift;
   my $app  = shift;
   my $conf = shift;

   $self->app( $app ) ;

   $self->set_conf($conf);
   $self->add_plugin( "Mojolicious::Plugin::Prototype::MooseForm::ClassResolver" );
   $self->add_plugin( "Mojolicious::Plugin::Prototype::MooseForm::TemplateData" );
   my $pl_self = $self;

   $app->helper(prototype_conf => sub{
      my $self = shift;
      my $key  = shift;

      $key = "prototype_" . $key unless $key =~ s/^\+//;
      if(@_) {
         $pl_self->conf->{$key} = shift;
      } else {
         return $pl_self->conf->{$key};
      }
   });

   $app->helper("moose_form_get_conf" => sub {
      $pl_self->conf;
   });

   $app->helper("moose_form_template_for" => sub {
      my $self   = shift;
      my $action = lc shift;
      my $type   = lc shift;
      my $value  = lc shift;

      my $controller = $self;
      my $renderer   = $self->app->renderer;

      my @val = $pl_self->separate_value($value);
      my @ret;
      for(reverse 0 .. $#val) { 
         my $name = $pl_self->create_template_name($action, $type, @val[ 0 .. $_ ]);
         if(eval { $renderer->render($controller, { template => $name, subtype => $pl_self->type_reconstruct( @val[ $_ + 1 .. $#val ] ) }) } or $@) { 
            push @ret, $name, subtype => $pl_self->type_reconstruct(@val[ $_ + 1 .. $#val ] );
            last;
         }
      }
      if( not @ret ) {
         my @other = ( $action, $type, "default" );
         push @ret, $pl_self->create_template_name(@other), subtype => undef
      }
      @ret
   });

   $app->helper("get_defaults" => sub {
      my $self  = shift;
      my $class = shift;
      my @defaults = $pl_self->exec(get_class_details => $class, $self->flash("params"));
      $self->stash->{attributes} = [ @defaults ];
      $self->stash->{class}      = $class;
      $self->stash->{error}      = $self->flash("error");
      $self->stash->{action}     = "";
      $self->flash(error => {})
   });

   $app->helper("create_object" => sub {
      my $self  = shift;
      my $class = shift;
      my $params = {};
      $params->{$_} = $self->param($_) for $self->param;

      my $obj = $pl_self->exec(create_obj => $class, $params);
      my $error = $pl_self->error;
      if(keys %$error) {
         $self->flash(params => $obj);
         for my $key(keys %$error) {
            $app->log->debug("ERROR: " . $error->{ $key }) if $error->{ $key } ;
         }
         $self->flash(error => $error);
         return;
      }
      return $obj
   });
   $app->routes->add_shortcut("get_moose_form" => sub{
      my $self  = shift;
      my $url   = shift;
      my $class = shift;

      my $code   = ( grep{ref eq "CODE"} @_ )[ 0 ] ;
      my $action = ( map{$_->{ action }} grep {ref eq "HASH" and exists $_->{action}} @_ )[ 0 ] ;

      $self->get($url, sub{
         my $self = shift;
         $self->get_defaults($class);
         $self->stash->{action} = $action if $action;
         $self->$code(@_) if defined $code;
      }, grep{ref ne "CODE"} @_);
   });
   $app->routes->add_shortcut("moose_form" => sub{
      my $self  = shift;
      my $url   = shift;
      my $class = shift;

      my $code   = ( grep{ref eq "CODE"} @_ )[ 0 ] ;
      my $action = ( map{$_->{ action }} grep {ref eq "HASH" and exists $_->{action}} @_ )[ 0 ] ;
      my ($pname, $gname) = reverse grep{not ref} @_;

      my $purl = $action || $url;

      $self->get_moose_form($url, $class, ( $gname || "moose_form" ) );
      $self->post($purl, sub{
         my $self = shift;
         $self->stash->{ obj } = $self->create_object($class);
         return $self->redirect_to($url) if not $self->stash->{ obj };
         $self->stash->{action} = $action;
         $self->$code($self->stash->{ obj }, @_) if defined $code;
      } => $pname => grep{ref and ref ne "CODE"} @_);
   });

   *main::get_moose_form = sub{$app->routes->get_moose_form(@_)};
   *main::moose_form     = sub{$app->routes->moose_form(@_)};

   $app->routes->get("/css/moose_form" => "moose_form");
}

42

__END__

=head1 NAME

Mojolicious::Plugin::Prototype::MooseForm - A Mojolicious Plugin to make it eazy to create Form prototype from Moose Classes.

=head1 VERSION

Version 0.01

=cut



=head1 SYNOPSIS

Quick summary of what the module does.

Perhaps a little code snippet.

    package My::Class;
    use Moose;
    
    has my_string               => (is => 'rw', isa => "Str",                  documentation => "This is a string and it can't be undefined.");
    has my_number               => (is => 'rw', isa => "Num",                  documentation => "This is a number and it can't be undefined.");
    has maybe_a_string          => (is => 'rw', isa => "Maybe[Str]",           documentation => "This is a string but can be undefined.");
    has maybe_a_number          => (is => 'rw', isa => "Maybe[Num]",           documentation => "This is a number but can be undefined.");
    has array_of_strings        => (is => 'rw', isa => "ArrayRef[Str]",        documentation => "This is a array of strings.");
    has array_of_numbers        => (is => 'rw', isa => "ArrayRef[Num]",        documentation => "This is a array of numbers.");
    has array_of_maybe_strings  => (is => 'rw', isa => "ArrayRef[Maybe[Str]]", documentation => "This is a array of strings.");
    has array_of_maybe_numbers  => (is => 'rw', isa => "ArrayRef[Maybe[Num]]", documentation => "This is a array of numbers.");

    
    package main;
    
    use Mojolicious::Lite;
    
    BEGIN{ plugin "Mojolicious::Plugin::Prototype::MooseForm" }
    
    moose_form "/" => "My::Class" => sub {
       my $self = shift;
       my $obj  = shift;

       $self->render_json($obj);
    } => "response";
    
    app->start;

=head1 EXPORT

A list of functions that can be exported.  You can delete this section
if you don't export anything, such as for a purely object-oriented module.

=head1 SUBROUTINES/METHODS

=head2 function1

=cut

=head2 function2

=cut

=head1 AUTHOR

Fernando Correa de Oliveira, C<< <fco at cpan.org> >>

=head1 BUGS

Please report any bugs or feature requests to C<bug-mojolicious-plugin-prototype-mooseform at rt.cpan.org>, or through
the web interface at L<http://rt.cpan.org/NoAuth/ReportBug.html?Queue=Mojolicious-Plugin-Prototype-MooseForm>.  I will be notified, and then you'll
automatically be notified of progress on your bug as I make changes.




=head1 SUPPORT

You can find documentation for this module with the perldoc command.

    perldoc Mojolicious::Plugin::Prototype::MooseForm


You can also look for information at:

=over 4

=item * RT: CPAN's request tracker (report bugs here)

L<http://rt.cpan.org/NoAuth/Bugs.html?Dist=Mojolicious-Plugin-Prototype-MooseForm>

=item * AnnoCPAN: Annotated CPAN documentation

L<http://annocpan.org/dist/Mojolicious-Plugin-Prototype-MooseForm>

=item * CPAN Ratings

L<http://cpanratings.perl.org/d/Mojolicious-Plugin-Prototype-MooseForm>

=item * Search CPAN

L<http://search.cpan.org/dist/Mojolicious-Plugin-Prototype-MooseForm/>

=back


=head1 ACKNOWLEDGEMENTS


=head1 LICENSE AND COPYRIGHT

Copyright 2012 Fernando Correa de Oliveira.

This program is free software; you can redistribute it and/or modify it
under the terms of either: the GNU General Public License as published
by the Free Software Foundation; or the Artistic License.

See http://dev.perl.org/licenses/ for more information.


=cut

1; # End of Mojolicious::Plugin::Prototype::MooseForm
